src/seraphis_wallet/key_container.cpp

```diff
@@ -18,170 +46,217 @@
-KeyContainer::KeyContainer(jamtis_keys &&keys, const crypto::chacha_key &key)
-    : m_keys{keys}, m_encrypted{false}, m_encryption_iv{}
-{
-    encrypt(key);
-}
-//-------------------------------------------------------------------------------------------------------------------
-KeyContainer::KeyContainer(jamtis_keys &&keys, bool encrypted, const crypto::chacha_iv encryption_iv)
-    : m_keys{keys}, m_encrypted{encrypted}, m_encryption_iv{encryption_iv}
-{
+    // 3. recover jamtis keys 
+    JamtisKeys recovered_keys{};
+    recover_jamtis_keys(ser_keys, recovered_keys);
+
+    // 4. check if keys are valid and move to m_keys if so
+    CHECK_AND_ASSERT_THROW_MES(jamtis_keys_valid(recovered_keys, chacha_key), "load_from_keys_file: failed validating jamtis keys.");
+    m_keys = std::move(recovered_keys);
+
+    // 5. encrypt keys in memory
+    encrypt(chacha_key);
+
+    return true;
 }
 //-------------------------------------------------------------------------------------------------------------------
-boost::optional<KeyContainer> KeyContainer::load_from(const std::string &path, const crypto::chacha_key &key)
+bool KeyContainer::jamtis_keys_valid(const JamtisKeys &keys, const crypto::chacha_key &chacha_key)
 {
-    ser_KeyContainer ser_key_container;
+    // 1. make test_keys = keys
+    JamtisKeys test_keys{keys};
 
-    if (!read_encrypted_file(path, key, ser_key_container))
-        return {};
-
-    const KeyContainer key_container = ser_container_to_usable(ser_key_container, key);
-
-    if (!jamtis_keys_valid(key_container.m_keys))
+    // 2. derive keys
+    switch (get_wallet_type())
     {
-        return boost::none;
+        case WalletType::Master:
+        {
+            sp::jamtis::make_jamtis_unlockamounts_key(test_keys.k_vb, test_keys.xk_ua);
+            sp::jamtis::make_jamtis_findreceived_key(test_keys.k_vb, test_keys.xk_fr);
+            sp::jamtis::make_jamtis_generateaddress_secret(test_keys.k_vb, test_keys.s_ga);
+            sp::jamtis::make_jamtis_ciphertag_secret(test_keys.s_ga, test_keys.s_ct);
+            sp::make_seraphis_spendkey(test_keys.k_vb, test_keys.k_m, test_keys.K_1_base);
+            sp::jamtis::make_jamtis_unlockamounts_pubkey(test_keys.xk_ua, test_keys.xK_ua);
+            sp::jamtis::make_jamtis_findreceived_pubkey(test_keys.xk_fr, test_keys.xK_ua, test_keys.xK_fr);
+            break;
+        }
+        case WalletType::ViewOnly:
+        {
+            sp::jamtis::make_jamtis_findreceived_pubkey(test_keys.xk_fr, test_keys.xK_ua, test_keys.xK_fr);
+            break;
+        }
+        case WalletType::ViewBalance:
+        {
+            sp::jamtis::make_jamtis_findreceived_key(test_keys.k_vb, test_keys.xk_fr);
+            break;
+        }
+        default:
+        {
+            return false;
+            break;
+        }
     }
 
-    return key_container;
+    // 3. check if derived keys are correct
+    return test_keys == keys;
```

Comment:
Just a note, this assumes here that the uninitialized keys are valid; I think this is a fair assumption